python/pyarrow/tests/test_table.py

```diff
@@ -239,23 +239,65 @@ def test_concat_tables():
 
 
 def test_table_pandas():
-    data = [
-        pa.array(range(5)),
-        pa.array([-10, -5, 0, 5, 10])
-    ]
-    table = pa.Table.from_arrays(data, names=('a', 'b'))
-
-    # TODO: Use this part once from_pandas is implemented
-    # data = {'a': range(5), 'b': [-10, -5, 0, 5, 10]}
-    # df = pd.DataFrame(data)
-    # pa.Table.from_pandas(df)
+    data = {'a': range(5), 'b': [-10, -5, 0, 5, 10]}
+    df = pd.DataFrame(data)
+    table = pa.Table.from_pandas(df)
 
     df = table.to_pandas()
     assert set(df.columns) == set(('a', 'b'))
     assert df.shape == (5, 2)
     assert df.loc[0, 'b'] == -10
 
 
+def test_table_pandas_with_schema():
+    data = OrderedDict({
+        'a': range(5),
+        'b': [-10, -5, 0, 5, 10],
+        'c': np.array([-10, -5, 0, 5, 10], dtype=np.int32)
+    })
+    df = pd.DataFrame(data)
+
+    schema = pa.schema([
+        pa.field('a', pa.int64()),
+        pa.field('b', pa.int64()),
+        pa.field('c', pa.int32())
+    ])
+    table = pa.Table.from_pandas(df, schema=schema)
+    assert table.column(0).type == pa.int64()
+    assert table.column(1).type == pa.int64()
+    assert table.column(2).type == pa.int32()
+
+    df = table.to_pandas()
+    assert all(df.dtypes == [np.int64, np.int64, np.int32])
+    assert set(df.columns) == set(('a', 'b', 'c'))
+    assert df.shape == (5, 3)
+    assert df.loc[0, 'b'] == -10
+
+
+def test_table_pandas_with_partial_schema():
+    data = OrderedDict([
+        ('a', range(5)),
+        ('b', np.array([-10, -5, 0, 5, 10], dtype=np.int32)),
+        ('c', [-10, -5, 0, 5, 10])
+    ])
+    df = pd.DataFrame(data)
+
+    schema = pa.schema([
+        pa.field('a', pa.int64()),
+        pa.field('b', pa.int32())
+    ])
```

Comment:
This was the behaviour we had pre-0.4.0. Currently, the main usecase for the schema is to act as a type-indicator for columns where we cannot automatically infer its Arrow representation.